### PR TITLE
Add ?rankup command

### PR DESCRIPTION
**Pull request by Kacarott**
Added `?rankup` command, which can be used by users to display how many more katas of specific ranks they must complete, to rank up. (or, to overtake a specified user).

Supports various named arguments `--language` `--target` `--mode` `--limit`. A specific username can be provided. If not provided, then the Discord server nickname of the user invoking the command is used. All data comes from the standard CW api.

**Purpose:** This is meant merely as a fun "toy" and to increase competition and motivation in users. Reception to screenshots of the development have been received very well in the discord server already.

Example usage:
---
Kacarott: ?rankup
Codewars: Kacarott needs to complete:
```
22 1kyu
25 2kyu
27 3kyu
31 4kyu
34 5kyu
39 6kyu
41 7kyu
43 8kyu
```
to reach `2dan`

---
Kacarott: ?rankup --language=python --target=Blind4Basics --mode=each
Codewars: Kacarott needs to complete:
```
   45    1kyu
or 122   2kyu
or 331   3kyu
or 896   4kyu
or 2347  5kyu
or 6160  6kyu
or 16426 7kyu
or 24639 8kyu
```
to overtake Blind4Basics in python

### CODE REVIEW

File: src/events/message/commands/fun/rankup.ts
Comment (kazk) [inline src/events/message/commands/fun/rankup.ts:30]: I usually avoid `enum` because the equivalent doesn't exist in JavaScript, and I forget how it's represented there.

I usually do:
```typescript
const LEAST = "least";
const EACH = "each";
const SPREAD = "spread";
type Mode = LEAST | EACH | SPREAD;
```
And define a [type predicate](https://www.typescriptlang.org/docs/handbook/advanced-types.html#user-defined-type-guards):
```typescript
function isMode(value: unknown): value is Mode {
  return typeof value == "string" && (value == LEAST || value == EACH || value == SPREAD);
  // Or define `MODES` and do `MODES.indludes(value)`
}
```
This will make sure the value is `Mode`.

File: src/events/message/commands/fun/rankup.ts
Comment (kazk) [inline src/events/message/commands/fun/rankup.ts:54]: The formula should be `floor(exp(rank)*12 - 12)` where rank is `0..=15` for 8 kyu to 8 dan.

File: src/events/message/commands/fun/rankup.ts
Comment (kazk) [inline src/events/message/commands/fun/rankup.ts:149]: I'd extract this to a separate function that returns `[nextRank: string, remaining: number]`. The error cases can be done by throwing an error with message used by `send`, or changing the return type to include error message string (`[nextRank: string, remaining: number] | string`).

File: src/events/message/commands/fun/rankup.ts
Comment (kazk) [inline src/events/message/commands/fun/rankup.ts:168]: I'd extract formatting to a separate function.

**Review by kazk — CHANGES_REQUESTED** [review]

**Comment by kazk** [comment]
By the way, you should create a new branch for each PR. Delete the branch once it's merged, and sync your main with the upstream (this repo). This PR contains commits from the previous PR.

File: package-lock.json
Comment (kazk) [inline package-lock.json]: What version of Node.js did you use? I'd rather not have the lockfile updated like this in this PR.

File: src/events/message/commands/fun/rankup.ts
Comment (kazk) [inline src/events/message/commands/fun/rankup.ts:91]: This sucks, and I'm pretty sure it's a bug. But CW API v1 has been like this as long as I can remember, so fixing this can break some code.

Comment (kazk) [reply src/events/message/commands/fun/rankup.ts:91]: Wait, the API does respond with status code `404`. I just misread this.

File: src/events/message/commands/fun/rankup.ts
Comment (kazk) [inline src/events/message/commands/fun/rankup.ts:98]: You can add `Options` type (as long as values are string), and change the type of `opts`:

```typescript
export type Options = {
  // note that these are optional with `?`
  mode?: string;
  language?: string;
  target?: string;
  limit?: string;
};

export default async (message: Message, args: CommandArg[], opts: Options) => {
```

This is better because it forces you to handle the undefined cases correctly.

For example, the code in this PR is not affected, but `RegExp.test` stringifies the argument, so
```javascript
const opts = {};
/un/.test(opts.foo); //=> true
```

File: src/events/message/commands/fun/rankup.ts
Comment (Kacarott) [reply src/events/message/commands/fun/rankup.ts]: I've now replaced this with a full list of language IDs, so I can properly validate the `--language` argument.

File: src/events/message/commands/fun/rankup.ts
Comment (hobovsky) [inline src/events/message/commands/fun/rankup.ts:159]: I would suggest checking correctness of requested language only if the requested language does not exist in data returned for the user(s). This way we avoid necessity of updating the list here immediately after a new language is added. After all, if there is data returned for a language, then its valid regardless of being present on the list.

Comment (kazk) [reply src/events/message/commands/fun/rankup.ts:159]: Yeah, that should be good enough. We don't need to maintain the list here.

File: src/events/message/commands/fun/rankup.ts
Comment (kazk) [inline src/events/message/commands/fun/rankup.ts:128]: I'd define types for the API response and validate the structure, but I'll leave that up to you. It might be too much when you're just learning TypeScript. We can always add the types later.

If you want to try, I'd do something like the following:
```typescript
const getUserLangScore = async (user: string, lang: string = "overall"): Promise<number | undefined> => {
  const response = await fetch("https://www.codewars.com/api/v1/users/" + user);
  if (response.status === 404) return;
  const result = await response.json();
  if (!isRanksResponse(result)) return;
  if (lang !== "overall") {
    // TODO validate `lang`
    return result.ranks.languages[lang].score;
  }
  return result.ranks.overall.score;
};

type RanksResponse = {
  success: true;
  ranks: {
    overall: RankInfo;
    languages: { [lang: string]: RankInfo };
  };
};
type RankInfo = {
  rank: number;
  score: number;
};
const isRanksResponse = (json: unknown): json is RanksResponse => {
  //
};
```

File: src/events/message/commands/fun/rankup.ts
Comment (Kacarott) [reply src/events/message/commands/fun/rankup.ts:159]: I am not sure I understand how this would help. Either way, the list of languages would need to be here and updated, wouldn't it? Because sometimes a user might query a language which exists, but that they haven't started.

Comment (kazk) [reply src/events/message/commands/fun/rankup.ts:159]: I don't remember what I was thinking 6 hours ago. I guess I wasn't considering querying the language the user haven't started yet.

Maintaining the list is not a big deal (I'm sure one of you will update), so I'm OK with this.

Comment (Kacarott) [reply src/events/message/commands/fun/rankup.ts:159]: The other option (how it was handled prior to my last push) is to just show a generic `"${username} has not started training ${language}, or the language ID is wrong"` with a link to the docs maybe. Then there's no need to keep an updated list.

File: src/events/message/commands/fun/rankup.ts
Comment (Kacarott) [reply src/events/message/commands/fun/rankup.ts:128]: Yeah I'll give that a go, I should be able to manage, I think. But I might refactor out the API request to a new file that exports a single function `getCWUserInfo` (or something) and also handles all its own relevant types, if that sounds ok?

File: src/events/message/commands/fun/rankup.ts
Comment (kazk) [reply src/events/message/commands/fun/rankup.ts:159]: Yeah, that's fine too.

```
Kacarott has not started training `brainfuck`, or the language ID is invalid.
See <https://docs.codewars.com/languages/> to find the correct ID.
```

File: src/events/message/commands/fun/rankup.ts
Comment (kazk) [reply src/events/message/commands/fun/rankup.ts:128]: Yeah, sounds good to me. I'd name it `getUser` in a `codewars` module.

I forgot that [`zod`](https://github.com/colinhacks/zod) is already installed. You can use that to create types with validation.

**Comment by Kacarott** [comment]
The requested changes should be implemented. I'm happy to try and fix the package-lock thing, if you can tell me exactly what steps to take.

File: src/events/message/commands/fun/rankup.ts
Comment (Kacarott) [reply src/events/message/commands/fun/rankup.ts:159]: My logic was, the user was not found and therefore cannot progress past 8kyu. I think using callers name would be confusing, implying it checked CW for their name instead of the one provided.

I can also remove this and provide a more "normal" message, if that would be better.

**Comment by kazk** [comment]
If you want to fix the lock file, you can start interactive rebase, edit the first commit, reset the changes, restore the original lock file, install the packages again with NPM v6 which will update the lock file without changing the version. Then commit the new changes, and continue rebasing to the end.

A more hacky way is to copy the lock file from this repo and use NPM v6 to install and update that lock file.

Like I said in Discord, it's okay if you can't fix it.

File: src/events/message/commands/fun/rankup.ts
Comment (kazk) [reply src/events/message/commands/fun/rankup.ts:159]: Huh, I thought the message was referring to the mistake the caller made. Like you (the caller) can't even type the username correctly.

Yeah, I prefer a normal message.

**Comment by Kacarott** [comment]
I believe I managed to fix it (along with other requested changes)

**Review by kazk — APPROVED** [review]
Looks good to me!

**Pull request merged**
